### ISSUE TITLE
CPU intensive render in FF (not Chrome)

### ISSUE THREAD

**Issue by rade** [weaveworks/scope#2302]
When viewing [Weave Cloud Prod in prod, services, all namespaces, graph mode](https://cloud.weave.works/app/loud-breeze-77#!/state/{%22controlPipe%22:null,%22topologyViewMode%22:%22topo%22,%22nodeDetails%22:[],%22pinnedMetricType%22:null,%22pinnedSearches%22:[],%22searchQuery%22:%22%22,%22selectedNodeId%22:null,%22gridSortedBy%22:null,%22gridSortedDesc%22:null,%22topologyId%22:%22services%22,%22topologyOptions%22:{%22processes%22:{%22unconnected%22:%22hide%22},%22containers%22:{%22system%22:%22application%22,%22stopped%22:%22running%22,%22pseudo%22:%22hide%22},%22pods%22:{%22namespace%22:%22all%22,%22pseudo%22:%22hide%22}},%22contrastMode%22:false}), firefox uses an entire core on my machine. By contrast, Chrome uses less than 5%.

**Comment by rade** [comment]
I wonder whether #2291 would help here. Profiling tells me that most of the time is going on lots of "Recalculate Style" with "Restyle Hint: CSSAnimations".

**Comment by rade** [comment]
> I wonder whether #2291 would help here.

Apparently not. It's running on dev now but I am still seeing the same behaviour on [prod-in-dev](https://frontend.dev.weave.works/app/nameless-night-79#!/state/{%22controlPipe%22:null,%22topologyViewMode%22:%22topo%22,%22nodeDetails%22:[],%22pinnedMetricType%22:null,%22pinnedSearches%22:[],%22searchQuery%22:%22%22,%22selectedNodeId%22:null,%22gridSortedBy%22:%22label%22,%22gridSortedDesc%22:null,%22topologyId%22:%22services%22,%22topologyOptions%22:{%22processes%22:{%22unconnected%22:%22hide%22},%22containers%22:{%22system%22:%22all%22,%22stopped%22:%22running%22,%22pseudo%22:%22hide%22},%22pods%22:{%22namespace%22:%22all%22,%22pseudo%22:%22hide%22}},%22contrastMode%22:false}).

**Comment by fbarl** [comment]
I need to investigate this more but so far it seems that `Recalculate style` events are the bottleneck in FF. They are mostly triggered either by CSS animations/transitions or [react-motion](https://github.com/chenglou/react-motion) - disabling those puts down the CPU usage in the latent state close to 0%, so I feel we're close to pinning down the problem.

**Comment by fbarl** [comment]
@rade Curious whether you find the new FF performance good enough to close this issue :)

**Comment by rade** [comment]
It's fine now - using 25% of a core instead of 100%.

I do find the animation when opening/closing the details panel a bit too rapid now, making it difficult to keep a rough visual track of the nodes involved (which, as I understand it, is the purpose of the animation).

**Comment by fbarl** [comment]
> I do find the animation when opening/closing the details panel a bit too rapid now...

Funny enough, but making the animations faster contributed a lot to reducing the CPU load on Firefox.

The reason is that CSS style recalculation apparently eats up a lot of CPU on FF and since it would be triggered constantly for node/edge motions every 5 seconds (which is the current refresh interval via websockets), the refresh animation duration would determine how much time the CPU would appear very busy.

It used to be that the slow animation would take about 4 seconds to complete, meaning something would be animating 4 out of 5 seconds - roughly speaking, by making animations 4x faster, this has dropped to 1 out of 5 seconds, reducing the CPU load for animations almost 4x.

**Comment by rade** [comment]
> Funny enough, but making the animations faster contributed a lot to reducing the CPU load on Firefox.

I was aware of that.

> The reason is that CSS style recalculation apparently eats up a lot of CPU on FF and since it would be triggered constantly for node/edge motions every 5 seconds (which is the current refresh interval via websockets)

I don't understand. When nothing changes during a refresh, why are there animations happening? Also, if that is somehow unavoidable, could we somehow separate these animations from the user triggered animations (e.g. when opening the details panel), and slow down the latter but not the former?

**Comment by fbarl** [comment]
> When nothing changes during a refresh, why are there animations happening?

The case I actually had in mind is when there is a focused node in the circular layout, with a lot of connections that keep changing often (so some nodes going in and out in every update). It's not that hard to find such a case on our _dev_ and I think it's also an important use case.

> Also, if that is somehow unavoidable, could we somehow separate these animations from the user triggered animations (e.g. when opening the details panel), and slow down the latter but not the former?

This actually wouldn't be that easy to do and I think it's also not that great for the UI consistency to split into different cases there.

I think things would still work fine if I made the animation 2x faster (which would still be 2x slower than how it used to be) - I created #2803 for you to take a look at @rade.
